### DISCUSSION TITLE
Dependency Injection, IoC Container

### DISCUSSION THREAD

**Discussion by Pingear** [serverpod/serverpod#670] [Q&A]
Hello)
I did not find information about the IoC container in the documentation, how is dpendecy injection implemented in the serverpod?

**Comment by vlidholt** [comment]
Can you describe the use case and what you need in a bit more detail and with some example?

**Reply by Pingear** [reply]
Hello) I probably won’t be able to properly support the discussion about inversion of control, but in short, in order not to create a bunch of boilerplate in the form of interfaces that help implement inversion of control, Ioc controllers are usually built into servers. It's about the last letter in SOLID D - dependensy inversion. Commonly used decorators are -
<img width="1265" alt="image" src="https://user-images.githubusercontent.com/16469484/219901202-b3ad8b30-c983-4a7b-a33f-bf59da157fb5.png">


With this approach, the code is much cleaner. I've seen this in JavaEE and NestJS.

**Reply by Pingear** [reply]
And angular also)

**Reply by II11II** [reply]
Inversion of Control is needed for loose coupling between components in a system. It's important because it helps make systems more modular, flexible, and easier to maintain over time.

For business(or complex-logically) products, components often have a lot of direct dependencies on other components in the system. For example, if Component A needs to use Component B, it would create an instance of Component B directly within its code. This tight coupling can make it difficult to modify or test the system as a whole since changes to one component can easily affect others.

Writing the whole business logic inside serverpod's endpoint file is a bad idea. The reason for it is,

1. Tight coupling between components: Without Dependency Injection, components in a system are often tightly coupled to each other. This makes it difficult to replace or modify one component without affecting the others, and can make it harder to test the system as a whole.

2. Hard to test: When components are tightly coupled, it can be difficult to isolate them for testing. This can lead to more complex and time-consuming testing procedures, and can make it harder to catch bugs and issues before they reach production.

3. Increased complexity: When components are not decoupled through Dependency Injection, it can lead to more complex and harder-to-understand code. This can make it more difficult to maintain the code over time, and can make it harder for other developers to understand and work with the code.

4. Lack of flexibility: Without Dependency Injection, it can be harder to modify or extend a system over time. This can limit the ability of the system to adapt to changing requirements or business needs.

5. Code duplication: Without Dependency Injection, it's common for code to be duplicated across different components in a system. This can lead to more code to maintain and increase the risk of errors.

Overall, not using Dependency Injection can lead to a less maintainable, less flexible, and harder-to-test software system. By decoupling components through Dependency Injection, software developers can make their systems more modular, easier to test, and more adaptable to changing requirements over time.

**Reply by II11II** [reply]
@vlidholt

**Reply by II11II** [reply]
@Pingear For now, Serverpod does not support passing dependencies by constructor. But for now, you can use ```get_it``` package to implement DI

**Comment by migalv** [comment]
Hey @vlidholt - Any updates on this? I think this can be important to ensure good quality code. Maybe you or someone in the community has found a way around this.

**Comment by migalv** [comment]
Hey @SandPod any idea about this?

**Reply by SandPod** [reply]
@migalv - Sorry for the nonresponsiveness in this discussion thread.

I hope to get back to you with some input on the discussion during the week.

**Reply by migalv** [reply]
That would be very nice @SandPod 

Any updates?

**Comment by SandPod** [comment]
There is an ongoing discussion and issue around this here: https://github.com/serverpod/serverpod/issues/3154, with a suggested implementation that was closed here: https://github.com/serverpod/serverpod/pull/3155

Serverpod doesn't provide native support for dependency injection. But this is something we would like to introduce. This is already required when developing modules and would greatly help with configuring services for modules.

**Comment by migalv** [comment]
I see, thanks for the update @SandPod

**Reply by SandPod** [reply]
Service locator is being implemented here now: https://github.com/serverpod/serverpod/pull/3617

**Comment by obiwanzenobi** [comment]
@migalv you can try our DI library luanchpad: https://pub.dev/packages/launchpad
It supports local endpoint scope and server scope. 
Works well with getit and injectable
